### PR TITLE
[Integration][PagerDuty] - Export User Emails

### PR DESCRIPTION
**Pull request by PeyGis**
# Description

What - Updated the Assignees and Users property from string to email
Why - Currently the schedule and incident resources only contain reference  pagerduty user id when referencing users. we would like to have a way to map userId to user email in port so eventually any related resource/incident of specific user will be searchable by email.
How - Query all the users from the PagerDuty users API, store them in a cache, then fetch from the cache for every schedule  user.

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

## Screenshots

Incident Assignees
Screenshot 2024-05-23 at 4 09 26 PM

Schedule User
Screenshot 2024-05-23 at 4 21 27 PM

## API Documentation

[PagerDuty Users API](https://developer.pagerduty.com/api-reference/c96e889522dd6-list-users)

### CODE REVIEW

File: integrations/pagerduty/.port/resources/blueprints.json [hunk 1]
```diff
@@ -126,9 +126,12 @@
           "title": "Description"
         },
         "responder": {
-          "type": "string",
-          "title": "Assignee",
-          "format": "user"
+          "title": "Assignees",
```
Comment (Tankilevitch) [inline integrations/pagerduty/.port/resources/blueprints.json:129]: Let's update the identifier as well

File: integrations/pagerduty/clients/pagerduty.py [hunk 2]
```diff
@@ -253,3 +256,25 @@ async def get_service_analytics(
...
             logger.error(f"HTTP occurred while fetching service analytics data: {e}")
             return {}
+
+    async def fetch_and_cache_users(self) -> None:
+        async for users in self.paginate_request_to_pager_duty(data_key=USER_KEY):
+            for user in users:
+                event.attributes[user["id"]] = user
+
+    def get_cached_user(self, user_id: str) -> dict[str, Any] | None:
+        return event.attributes.get(f"{user_id}")
```
Comment (Tankilevitch) [inline integrations/pagerduty/clients/pagerduty.py:266]:
```suggestion
        return event.attributes.get(user_id})
```

File: integrations/pagerduty/clients/pagerduty.py [hunk 3]
```diff
@@ -253,3 +256,25 @@ async def get_service_analytics(
...
+        self, schedules: list[dict[str, Any]]
+    ) -> list[dict[str, Any]]:
+        await self.fetch_and_cache_users()
+
+        for schedule in schedules:
+            for user in schedule.get(USER_KEY, []):
+                cached_user = self.get_cached_user(user["id"])
+                if cached_user:
+                    user["__email"] = cached_user["email"]
+                else:
```
Comment (Tankilevitch) [inline integrations/pagerduty/clients/pagerduty.py:278]: What does the user object contains?

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/pagerduty/clients/pagerduty.py (hunk 3 above)
Comment (PeyGis) [reply integrations/pagerduty/clients/pagerduty.py:278]: Here's a sample user object

```json
[
       {
            "name": "User Name",
            "email": "user@getport.io",
            "time_zone": "Asia/Jerusalem",
            "color": "purple",
            "avatar_url": "https://gravatar.com/avatar/45c8178b572b583254694ab0aa3aaf4b.png?d=mm&r=PG",
            "billed": true,
            "role": "owner",
            "description": null,
            "invitation_sent": false,
            "job_title": null,
            "teams": [],
            "contact_methods": [
                {
                    "id": "PCQ9T3Z",
                    "type": "email_contact_method_reference",
                    "summary": "Default",
                    "self": "https://api.pagerduty.com/users/P4K4DLP/contact_methods/PCQ9T3Z",
                    "html_url": null
                }
            ],
            "notification_rules": [
                {
                    "id": "P87YHH4",
                    "type": "assignment_notification_rule_reference",
                    "summary": "0 minutes: channel PCQ9T3Z",
                    "self": "https://api.pagerduty.com/users/P4K4DLP/notification_rules/P87YHH4",
                    "html_url": null
                },
                {
                    "id": "POGJIWU",
                    "type": "assignment_notification_rule_reference",
                    "summary": "0 minutes: channel PCQ9T3Z",
                    "self": "https://api.pagerduty.com/users/P4K4DLP/notification_rules/POGJIWU",
                    "html_url": null
                }
            ],
            "id": "P4K4DLP",
            "type": "user",
            "summary": "User Name",
            "self": "https://api.pagerduty.com/users/P4K4DLP",
            "html_url": "https://domain-io.pagerduty.com/users/P4K4DLP"
        },...]
```

**Review by yairsimantov20 — APPROVED** [review]

**Review by Tankilevitch — APPROVED** [review]

**Pull request merged**